### ISSUE TITLE
Denied `clippy::non_ascii_literal` cannot be re-allowed for a single test function

### ISSUE THREAD

**Issue by TonalidadeHidrica** [rust-lang/rust-clippy#8263]
### Summary

Suppose that `clippy::non_ascii_literal` is denied globally.  Then I cannot allow it in a single `#[test]` function locally.

### Reproducer

I tried this code:

```rust
#![deny(clippy::non_ascii_literal)]

#[test]
#[allow(clippy::non_ascii_literal)]
pub fn it_works() {
    let _ = "悲しいかな、ここに日本語を書くことはできない。";
}
```

No clippy lint should occur, but a `clippy::non_ascii_literal` is detected.

If I removed `#[test]`, it is allowed as expected.  I suspect that this is due to `test` proc macro producing a code *outside* the testing function.

This issue may be related to, or can be fixed at the same time as, #7739.

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-apple-darwin
release: 1.57.0
LLVM version: 13.0.0
```
